### PR TITLE
Implement a coalescent time distribution

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #2426 
Blocking #2468 

This adds two new distributions `CoalescentTimes` and `CoalescentTimesWithRate` and a class `CoalescentRateLikelihood` implementing a phylogenetic likelihood
```
p(coalescent times | sample times, base rate)
```
by splitting a phylogeny into fixed sample times (i.e. EPI nodes, leaves) and unknown coalescent times (i.e. internal NODE nodes). For `...WithRates`, the base rate is specified as a vector representing a piecewise constant base coalescent rate on a uniform grid (e.g. computed from `S` and `I` time series in SIR models).

The design aims to make it cheap to enumerate over multiple phylogenetic samples (e.g. from BEAST). Two crucial observations are: (1) among phylogeny samples, the leaves are fixed, hence the total number of nodes is fixed; and (2) we can avoid summing over intervals by precomputing `rate.cumsum(-1)` and using this for all samples. Exploiting these facts leads to an `O(T + S N log(N))` algorithm where `T` is the number of simulation time steps, `N` is the number of leaves in the phylogeny, and `S` is the number of sampled phylogenies. This PR does not implement batching over different phylogenetic problems (e.g. one phylogeny per region in regional models); that would require more careful masking and padding and will be left to a possible future PR.

These are low level distribution and do not implement preprocessing to transform phylogenetic trees into `(coal_times, leaf_times)` pairs. This PR is intended as a low-level helper for use in #2468 and an example using `pyro.contrib.epidemiology`.

## Tested
- [x] shape tests
- [x] compare against uniform rate
- [x] added smoke tests for use in an SEIR model
- [x] visually sanity check samples in [a notebook](https://github.com/fritzo/notebooks/blob/master/coalescent.ipynb)

### CODE REVIEW

File: pyro/distributions/coalescent.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    @functools.wraps(fn)
+    def memoized_fn(*args):
+        key = tuple(map(id, args))
+        if key not in cache:
+            cache[key] = fn(*args)
+            for arg in args:
+                weakref.finalize(arg, cache.pop, key, None)
+        return cache[key]
+
+    return memoized_fn
```
Comment (fritzo) [inline pyro/distributions/coalescent.py:223]: @neerajprad could you review this code? I wonder if it is interacting poorly with PyTorch?

Comment (fritzo) [reply pyro/distributions/coalescent.py:223]: No the failure occurs even without this, so this can't be the problem...

**Comment by fritzo** [comment]
@eb8680 FYI I am adding an interface that plays better with `.transition_bwd()` as we discussed.

**Comment by fritzo** [comment]
I've added a `CoalescentRateLikelihood` class compatible with both vectorized and sequential operation, and copied more of your #2468 into this to exercise the likelihood in an SEIR model. However this PR still deals entirely with `(leaf_times, coal_times)`.

I think it would be cleanest to merge this PR, then update #2468 to use `CoalescentRateLikelihood` and update the example script there to parse a `.nwk` phylogeny and extract to `(leaf_times, coal_times)` either using your r2py code or the `Bio.Phylo` python library.

File: pyro/contrib/epidemiology/seir.py [hunk 2]
```diff
@@ -240,3 +262,8 @@ def transition_bwd(self, params, prev, curr, t):
         pyro.sample("obs_{}".format(t),
                     dist.ExtendedBinomial(S2E, rho),
                     obs=self.data[t])
+        if self.coal_likelihood is not None:
+            R = R0 * prev["S"] / self.population
+            coal_rate = R * (1. + 1. / k) / (tau_i * prev["I"] + 1e-8)
+            pyro.factor("coalescent_{}".format(t),
+                        self.coal_likelihood(coal_rate, t))
```
Comment (eb8680) [inline pyro/contrib/epidemiology/seir.py:269]: This seems like a good interface

File: pyro/distributions/coalescent.py [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+        i = coal_times.floor().clamp(min=0).long()
+        rates = phylogeny.coal_binomial * _gather(self.rate_grid, -1, i)
+        log_prob = log_prob + safe_log(rates).sum(-1)
+
+        batch_shape = broadcast_shape(self.batch_shape, value.shape[:-1])
+        log_prob = log_prob.expand(batch_shape)
+        return log_prob
+
+
+class CoalescentRateLikelihood:
```
Comment (eb8680) [inline pyro/distributions/coalescent.py:195]: Should this be a `nn.Module`?

Comment (fritzo) [reply pyro/distributions/coalescent.py:195]: At the moment there is no reason to make this an `nn.Module`. It is a little more similar to a `Distribution`, but doesn't quite follow the distribution interface (indeed I've given it a `validate_args` kwarg like a distribution). Moreover in typical usage the parameters are all fixed rather than learned.

BTW this is a very natural object from the perspective of Funsor. Consider a funsor
```py
leaf_times = Tensor(...)
coal_times = Tensor(...)
rate_grid = Variable("rate_grid", reals(duration))
f = CoalescentTimesWithRate(leaf_times, rate_grid).log_prob(coal_times)
```
Then this likelihood represents roughly the funsor
```py
f(rate_grid=rate_grid["t"])
```
I look forward to the bright future when we won't need special classes like this 🚀

**Comment by fritzo** [comment]
Thanks for reviewing!
